### ISSUE TITLE
Tensor::ShareData compilation error.

### ISSUE THREAD

**Issue by BrettRyland** [leonardvandriel/caffe2_cpp_tutorial#41]
When compiling with `make`, I get the following error:
```
/usr/bin/c++  -DWITH_CUDA -DWITH_GPU -DWITH_OPENCV -I/home/brett/Programming/github/caffe2_cpp_tutorial/include -I/usr/include/eigen3 -I/usr/include/opencv  -std=c++11 -fPIC    -std=gnu++11 -o CMakeFiles/caffe2_cpp.dir/src/caffe2/util/blob.cc.o -c /home/brett/Programming/github/caffe2_cpp_tutorial/src/caffe2/util/blob.cc
/home/brett/Programming/github/caffe2_cpp_tutorial/src/caffe2/util/blob.cc: In member function ‘void caffe2::BlobUtil::Set(const TensorCPU&, bool)’:
/home/brett/Programming/github/caffe2_cpp_tutorial/src/caffe2/util/blob.cc:24:28: error: no matching function for call to ‘caffe2::Tensor<caffe2::CUDAContext>::ShareData(const TensorCPU&)’
     tensor->ShareData(value);
                            ^
In file included from /home/brett/Programming/github/caffe2_cpp_tutorial/include/caffe2/util/blob.h:5:0,
                 from /home/brett/Programming/github/caffe2_cpp_tutorial/src/caffe2/util/blob.cc:1:
/usr/local/include/caffe2/core/tensor.h:416:8: note: candidate: void caffe2::Tensor<Context>::ShareData(const caffe2::Tensor<Context>&) [with Context = caffe2::CUDAContext]
   void ShareData(const Tensor& src) {
        ^~~~~~~~~
/usr/local/include/caffe2/core/tensor.h:416:8: note:   no known conversion for argument 1 from ‘const TensorCPU {aka const caffe2::Tensor<caffe2::CPUContext>}’ to ‘const caffe2::Tensor<caffe2::CUDAContext>&’
CMakeFiles/caffe2_cpp.dir/build.make:62: recipe for target 'CMakeFiles/caffe2_cpp.dir/src/caffe2/util/blob.cc.o' failed
```
I'm using the current caffe2 from github and here is my [CMakeCache](https://github.com/leonardvandriel/caffe2_cpp_tutorial/files/1660144/CMakeCache.txt). My system is Debian/unstable.

**Comment by mattdingmeng** [comment]
I got the same error. I built the latest caffe2.

**Comment by mattdingmeng** [comment]
@BrettRyland did you find a way to solve it? I think this tutorial is based on a old version of Caffe2. But I do not know which caffe2 version should I use. The latest version does not work.

**Comment by mattdingmeng** [comment]
@leonardvandriel which Caffe2 version you used for this tutorial? I got Tensor::ShareData error with the latest Caffe2.

**Comment by leonardvandriel** [comment]
This was probably introduced on Jan 19 in caffe2 commit adfb1f48: "Make tensor cast constructor explicit". I'll post a fix in bit.

**Comment by leonardvandriel** [comment]
Fixed in commit d77e101b5. Let me know if this works for you.

**Comment by mattdingmeng** [comment]
@leonardvandriel Awesome! It works now! Thank you!
